File: mobile/lib/protocol.ts
/**
 * protocol.ts — TypeScript mirror of internal/protocol/protocol.go
 *
 * Wire format:
 *   Text  frames → JSON Envelope
 *   Binary frames → BinaryFrame (1 byte kind | 2-byte big-endian idLen | id | payload)
 */

// ---------------------------------------------------------------------------
// Channel
// ---------------------------------------------------------------------------

/** Identifies which subsystem an Envelope targets. */
export type Channel = 'ctrl' | 'term' | 'fs' | 'sys' | 'api';

export const Ch = {
  Ctrl: 'ctrl' as Channel, // pairing, auth, ping/pong, session lifecycle
  Term: 'term' as Channel, // PTY terminal
  FS:   'fs'   as Channel, // file operations
  Sys:  'sys'  as Channel, // system metrics + process control
  API:  'api'  as Channel, // custom user-registered handlers
} as const;

// ---------------------------------------------------------------------------
// Envelope (text frames)
// ---------------------------------------------------------------------------

/**
 * JSON envelope sent over text WebSocket frames.
 * ID correlates a response with its request; omit for streaming/event messages.
 */
export interface Envelope {
  ch:    Channel;
  type:  string;
  id?:   string;
  /** Arbitrary JSON payload — keep as unknown and narrow at call sites. */
  data?: unknown;
  err?:  string;
}

// ---------------------------------------------------------------------------
// BinaryKind
// ---------------------------------------------------------------------------

/** Tags a binary frame so the receiver can route it. Matches Go BinaryKind. */
export enum BinaryKind {
  /** PTY output — StreamID = terminal session id */
  TermOutput = 1,
  /** PTY input  — StreamID = terminal session id */
  TermInput  = 2,
  /** File bytes — StreamID = transfer id */
  FSData     = 3,
}

// ---------------------------------------------------------------------------
// BinaryFrame helpers
// ---------------------------------------------------------------------------

export interface BinaryFrame {
  kind:     BinaryKind;
  streamID: string;
  payload:  Uint8Array;
}

const te = new TextEncoder();
const td = new TextDecoder();

/**
 * Encode a BinaryFrame to bytes for a WebSocket binary message.
 *
 * Layout (matches BinaryFrame.Encode in protocol.go):
 *   byte  0        : Kind
 *   bytes 1..2     : StreamID length (uint16, big-endian)
 *   bytes 3..3+n   : StreamID (UTF-8)
 *   remaining      : Payload
 */
export function encodeBinaryFrame(frame: BinaryFrame): Uint8Array {
  const idBytes = te.encode(frame.streamID);
  const idLen   = idBytes.byteLength;
  const buf     = new Uint8Array(3 + idLen + frame.payload.byteLength);
  const view    = new DataView(buf.buffer);

  buf[0] = frame.kind;
  view.setUint16(1, idLen, false /* big-endian */);
  buf.set(idBytes, 3);
  buf.set(frame.payload, 3 + idLen);

  return buf;
}

/**
 * Decode bytes from a WebSocket binary message into a BinaryFrame.
 * Throws if the frame is truncated (mirrors ErrShortFrame in protocol.go).
 */
export function decodeBinaryFrame(buf: ArrayBuffer | Uint8Array): BinaryFrame {
  const bytes = buf instanceof Uint8Array ? buf : new Uint8Array(buf);
  if (bytes.byteLength < 3) {
    throw new Error('protocol: short binary frame');
  }

  const view  = new DataView(bytes.buffer, bytes.byteOffset, bytes.byteLength);
  const kind  = bytes[0] as BinaryKind;
  const idLen = view.getUint16(1, false /* big-endian */);

  if (bytes.byteLength < 3 + idLen) {
    throw new Error('protocol: short binary frame');
  }

  const streamID = td.decode(bytes.subarray(3, 3 + idLen));
  const payload  = bytes.slice(3 + idLen); // copy, like Go's append([]byte(nil), ...)

  return { kind, streamID, payload };
}

File: internal/protocol/protocol.go
// Package protocol defines the wire format shared by hostd and the mobile app.
//
// A single WebSocket carries every channel. Two kinds of WS frames are used:
//
//   - Text frames  → JSON Envelope (control + request/response messages).
//   - Binary frames → BinaryFrame (high-throughput streams: terminal output,
//     file bytes) to avoid base64 overhead.
//
// The envelope is routed by its Ch (channel) field so new channels can be
// added without breaking older clients. See docs/PLAN.md for the design.
package protocol

import (
	"encoding/binary"
	"encoding/json"
	"errors"
)

// Channel identifies which subsystem an Envelope targets.
type Channel string

const (
	ChCtrl Channel = "ctrl" // pairing, auth, ping/pong, session lifecycle
	ChTerm Channel = "term" // PTY terminal
	ChFS   Channel = "fs"   // file operations
	ChSys  Channel = "sys"  // system metrics + process control
	ChAPI  Channel = "api"  // custom user-registered handlers
)

// Envelope is the JSON message sent over text WebSocket frames.
//
// ID correlates a response with its request (request/response channels such as
// fs and api). For streaming/event messages it may be empty.
type Envelope struct {
	Ch   Channel         `json:"ch"`
	Type string          `json:"type"`
	ID   string          `json:"id,omitempty"`
	Data json.RawMessage `json:"data,omitempty"`
	Err  string          `json:"err,omitempty"`
}

// NewEnvelope builds an envelope, JSON-encoding data into Data.
func NewEnvelope(ch Channel, typ, id string, data any) (Envelope, error) {
	e := Envelope{Ch: ch, Type: typ, ID: id}
	if data != nil {
		raw, err := json.Marshal(data)
		if err != nil {
			return Envelope{}, err
		}
		e.Data = raw
	}
	return e, nil
}

// ErrorEnvelope builds an error response envelope for a given request id.
func ErrorEnvelope(ch Channel, typ, id, msg string) Envelope {
	return Envelope{Ch: ch, Type: typ, ID: id, Err: msg}
}

// Bind unmarshals the envelope's Data into v.
func (e Envelope) Bind(v any) error {
	if len(e.Data) == 0 {
		return nil
	}
	return json.Unmarshal(e.Data, v)
}

// ---- Binary frames ----------------------------------------------------------

// BinaryKind tags a binary frame so the receiver can route it.
type BinaryKind byte

const (
	BinTermOutput BinaryKind = 1 // PTY output:   StreamID = terminal session id
	BinTermInput  BinaryKind = 2 // PTY input:    StreamID = terminal session id
	BinFSData     BinaryKind = 3 // file bytes:   StreamID = transfer id
)

// BinaryFrame is the layout for binary WebSocket frames:
//
//	byte 0        : Kind
//	bytes 1..2    : StreamID length (uint16, big-endian)
//	bytes 3..3+n  : StreamID (UTF-8)
//	remaining     : Payload
//
// StreamID names the logical stream (terminal session id, transfer id, …) so a
// single WebSocket can multiplex many concurrent binary streams.
type BinaryFrame struct {
	Kind     BinaryKind
	StreamID string
	Payload  []byte
}

// ErrShortFrame is returned when a binary frame is truncated.
var ErrShortFrame = errors.New("protocol: short binary frame")

// Encode serializes the frame to bytes for a WebSocket binary message.
func (f BinaryFrame) Encode() []byte {
	idLen := len(f.StreamID)
	buf := make([]byte, 3+idLen+len(f.Payload))
	buf[0] = byte(f.Kind)
	binary.BigEndian.PutUint16(buf[1:3], uint16(idLen))
	copy(buf[3:3+idLen], f.StreamID)
	copy(buf[3+idLen:], f.Payload)
	return buf
}

// DecodeBinaryFrame parses bytes from a WebSocket binary message.
func DecodeBinaryFrame(b []byte) (BinaryFrame, error) {
	if len(b) < 3 {
		return BinaryFrame{}, ErrShortFrame
	}
	idLen := int(binary.BigEndian.Uint16(b[1:3]))
	if len(b) < 3+idLen {
		return BinaryFrame{}, ErrShortFrame
	}
	return BinaryFrame{
		Kind:     BinaryKind(b[0]),
		StreamID: string(b[3 : 3+idLen]),
		Payload:  append([]byte(nil), b[3+idLen:]...),
	}, nil
}
